### PR TITLE
Rdm 10937

### PR DESCRIPTION
**Pull request by Thor-tech-of-metal**
RDM-10937: Change [CASE REFERENCE] metadata field type from 'Number' to 'Text'
               [RDM-10937] (https://tools.hmcts.net/jira/browse/RDM-10937).

### CODE REVIEW

File: Jenkinsfile_CNP [hunk 1]
```diff
@@ -116,7 +116,7 @@ withPipeline(type, product, component) {
             environmentOfDependencies = "aat"
         }
         env.DEFINITION_STORE_HOST      = "http://ccd-definition-store-api-${environmentOfDependencies}.service.core-compute-${environmentOfDependencies}.internal"
-        env.ELASTIC_SEARCH_FTA_ENABLED = "true"
```
Comment (Thor-tech-of-metal) [inline Jenkinsfile_CNP:119]: Elastic search befta has been enable for prs,  in dev branch.  The correct values has to be fixed in another pr, or this one . ☝️

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java [hunk 2]
```diff
@@ -11,4 +13,20 @@ public static String formatCaseReference(String caseReference) {
         }
         return caseReference;
     }
+
+    public static boolean checkRegex(final String value) {
+        final String regularExpression = "(?:^[0-9]{16}$|^\\d{4}-\\d{4}-\\d{4}-\\d{4}$)";
+        return value.matches(regularExpression);
+    }
+
+    public static Optional<String> getFormatCaseReference(Optional<String> caseReference) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:22]: why you passing the Optional? in terms of API it's better if you pass the String, and it's the code that calls the that deals with the Optional.
Otherwise every code that calls this method will need to wrap the string in an Optional

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java [hunk 3]
```diff
@@ -11,4 +13,20 @@ public static String formatCaseReference(String caseReference) {
         }
         return caseReference;
     }
+
+    public static boolean checkRegex(final String value) {
+        final String regularExpression = "(?:^[0-9]{16}$|^\\d{4}-\\d{4}-\\d{4}-\\d{4}$)";
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:18]: can we have this as a constant as a class field please. 
Also, can you give it a better name that helps understand what the regex is ?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java [hunk 4]
```diff
@@ -11,4 +13,20 @@ public static String formatCaseReference(String caseReference) {
...
+
+    public static Optional<String> getFormatCaseReference(Optional<String> caseReference) {
+
+        if (caseReference.isPresent()) {
+            if (!CaseReferenceUtils.checkRegex(caseReference.get())) {
+                return Optional.of("0000000000000000");
+            }
+            return Optional.of(CaseReferenceUtils.formatCaseReference(caseReference.get()));
+        }
+        return caseReference;
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:30]: I think all this logic cold be done by the calling class and this method be completely removed

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java [hunk 5]
```diff
@@ -11,4 +13,20 @@ public static String formatCaseReference(String caseReference) {
...
+        return value.matches(regularExpression);
+    }
+
+    public static Optional<String> getFormatCaseReference(Optional<String> caseReference) {
+
+        if (caseReference.isPresent()) {
+            if (!CaseReferenceUtils.checkRegex(caseReference.get())) {
+                return Optional.of("0000000000000000");
+            }
+            return Optional.of(CaseReferenceUtils.formatCaseReference(caseReference.get()));
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:28]: can you rename `formatCaseReference` to something more understandable? like for example: `removeHyphens`?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java [hunk 6]
```diff
@@ -11,4 +13,20 @@ public static String formatCaseReference(String caseReference) {
...
+    public static boolean checkRegex(final String value) {
+        final String regularExpression = "(?:^[0-9]{16}$|^\\d{4}-\\d{4}-\\d{4}-\\d{4}$)";
+        return value.matches(regularExpression);
+    }
+
+    public static Optional<String> getFormatCaseReference(Optional<String> caseReference) {
+
+        if (caseReference.isPresent()) {
+            if (!CaseReferenceUtils.checkRegex(caseReference.get())) {
+                return Optional.of("0000000000000000");
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:26]: is this safe?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java [hunk 7]
```diff
@@ -1,5 +1,7 @@
 package uk.gov.hmcts.ccd.domain.model.common;
 
+import java.util.Optional;
+
 public class CaseReferenceUtils {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:5]: some unit tests for this class would be useful

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/MergeDataToSearchResultOperation.java [hunk 8]
```diff
@@ -120,7 +120,7 @@ private SearchResultViewItem buildSearchResultViewItem(final CaseDetails caseDet
                                                            final CaseTypeDefinition caseTypeDefinition,
                                                            final SearchResultDefinition searchResult) {
         Map<String, JsonNode> caseData = new HashMap<>(caseDetails.getData());
-        Map<String, Object> caseMetadata = new HashMap<>(caseDetails.getMetadata());
+        Map<String, Object> caseMetadata = new HashMap<>(caseDetails.getFormattedMetadata());
```
Comment (rafalkalita) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/MergeDataToSearchResultOperation.java:123]: Change not reflected in unit test.

**Review by rafalkalita — COMMENTED** [review]
Few comments

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java [hunk 9]
```diff
@@ -330,6 +330,21 @@ private void setAfterSubmitCallbackResponseEntity(final AfterSubmitCallbackRespo
         return metadata;
     }
 
+    @JsonIgnore
+    public Map<String, Object> getFormattedMetadata() {
```
Comment (kiran-yenigala-hmcts) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java:334]: do we need completely new method for this, can't we change this in getMetaData method

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:26]: haha good point , we want the implementation to be like that ,  this represent the null in case_reference value. :)

Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:26]: I also wonder whether there's a better way to do this, so someone searches for a non valid reference, we don't execute the query and immediately return empty results. Your solution executes the query.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java (hunk 9 above)
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java:334]: probably not the best name. You are not formatting here. You are changing the type of one of the map entries

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java [hunk 10]
```diff
@@ -330,6 +330,21 @@ private void setAfterSubmitCallbackResponseEntity(final AfterSubmitCallbackRespo
...
+        if (metadata.isEmpty()) {
+            metadata.put(JURISDICTION.getReference(), getJurisdiction());
+            metadata.put(CASE_TYPE.getReference(), getCaseTypeId());
+            metadata.put(STATE.getReference(), getState());
+            metadata.put(CASE_REFERENCE.getReference(), getReference() != null ? getReference().toString() : getId());
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java:339]: for my understanding, can you explain where the code breaks if you don't do this conversion? maybe we can find a better way to do this

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:26]: we don't want an error @rafalkalita , what we want is that when the user searches with an invalid case reference, we want to just return no results

Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:26]: I think it would be better to add case reference validation here:
https://github.com/hmcts/ccd-data-store-api/blob/80adc76067063ba3c3600fb3e0674b41bfe5426f/src/main/java/uk/gov/hmcts/ccd/domain/service/search/DefaultSearchOperation.java#L26

so that if the user is searching by case reference and is an invalid one, we just return empty results without executing the query on the DB

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java (hunk 9 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java:334]: When getFormattedMetaData is called before getMetaData then it will populate metadata map already.
it may break other functionalities.

if you are confident that functionality is not going to break, then in both methods, only below line is different, 
            metadata.put(CASE_REFERENCE.getReference(), getReference() != null ? getReference() : getId());
            
            Which can be handled by boolean parameter, Think about this

**Review by kiran-yenigala-hmcts — COMMENTED** [review]
couple of comments

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java (hunk 10 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java:337]: Toolkit SearchResultViewItemComparatorFactory comparators. Browser errors

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java (hunk 9 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java:334]: What would be recommended name ?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/MergeDataToSearchResultOperation.java (hunk 8 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/MergeDataToSearchResultOperation.java]: There is not test in caseDetails  but I added a test  to evaluate the case reference type in MergeDataToSearchResultOperation.java

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java (hunk 9 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseDetails.java]: I ended up using getConvertedMetadata

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java (hunk 6 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java]: That one it is important !

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java [hunk 11]
```diff
@@ -2,13 +2,19 @@
 
 public class CaseReferenceUtils {
 
+    public static final String CASE_REFERENCE_EXPRESSION = "(?:^[0-9]{16}$|^\\d{4}-\\d{4}-\\d{4}-\\d{4}$)";
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:5]:
```suggestion
    public static final String CASE_REFERENCE_REGEX = "(?:^[0-9]{16}$|^\\d{4}-\\d{4}-\\d{4}-\\d{4}$)";
```

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java [hunk 12]
```diff
@@ -2,13 +2,19 @@
...
 
-    public static String formatCaseReference(String caseReference) {
+    public static String removeHyphens(String caseReference) {
         if (caseReference.contains("-")) {
             return String.join("", caseReference.split("-"));
         }
         return caseReference;
     }
+
+    public static boolean isAValidCaseReferenceFormat(final String value) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:17]:
```suggestion
    public static boolean isAValidCaseReference(final String reference) {
```

File: src/main/java/uk/gov/hmcts/ccd/domain/service/search/PaginatedSearchMetaDataOperation.java [hunk 13]
```diff
@@ -21,6 +20,10 @@ public PaginatedSearchMetaDataOperation(@Qualifier(CachedCaseDetailsRepository.Q
     }
 
     public PaginatedSearchMetadata execute(MetaData metaData, Map<String, String> criteria) {
+
+        if (!metaData.validateAndsConvertReference()) {
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/search/PaginatedSearchMetaDataOperation.java:24]:
```suggestion
        if (!metaData.validateAndConvertReference()) {
```

File: src/main/java/uk/gov/hmcts/ccd/domain/service/search/PaginatedSearchMetaDataOperation.java [hunk 14]
```diff
@@ -21,6 +20,10 @@ public PaginatedSearchMetaDataOperation(@Qualifier(CachedCaseDetailsRepository.Q
...
+            return new PaginatedSearchMetadata();
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/search/PaginatedSearchMetaDataOperation.java:25]:
```suggestion
            return PaginatedSearchMetadata.EMPTY;
```
where EMPTY is a constant in PaginatedSearchMetadata:
`public static final EMPTY = new PaginatedSearchMetadata();`

**Review by mario-paniccia — APPROVED** [review]
approved with some minor comments

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java (hunk 11 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:5]: DOne

File: src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java (hunk 12 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/common/CaseReferenceUtils.java:17]: DONE

File: src/main/java/uk/gov/hmcts/ccd/domain/service/search/PaginatedSearchMetaDataOperation.java (hunk 13 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/search/PaginatedSearchMetaDataOperation.java:24]: DONE

File: src/main/java/uk/gov/hmcts/ccd/domain/service/search/PaginatedSearchMetaDataOperation.java (hunk 14 above)
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/search/PaginatedSearchMetaDataOperation.java:25]: DONE

**Review by kiran-yenigala-hmcts — APPROVED** [review]

**Review by rafalkalita — APPROVED** [review]
